### PR TITLE
Storing sessions history

### PR DESCRIPTION
**Pull request by interro**
Updates #433 
- Uses BoltDB to store sessions
- Updating of session objects may be implemented in another PR to reduce scope

### CODE REVIEW

File: core/connection/manager.go [hunk 1]
```diff
@@ -155,6 +159,12 @@ func (manager *connectionManager) startConnection(consumerID, providerID identit
 		return err
 	}
 
+	providerCountry := "UNKNOWN" //proposal.ServiceDefinition.GetLocation().Country
+	err = manager.saveSession(sessionID, providerID, proposal.ServiceType, providerCountry)
```
[the remark this answers is not in the document]
Comment (zolia) [reply core/connection/manager.go]: yes, we have `ConnectOptions` or `ConnectParams` struct for that.

[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager.go]: done

File: core/connection/manager.go [hunk 2]
```diff
@@ -270,3 +280,21 @@ func (manager *connectionManager) onStateChanged(state State, sessionID session.
...
+	se := Session{
+		SessionID:       sessionID,
+		ProviderID:      providerID,
+		ServiceType:     serviceType,
+		ProviderCountry: providerCountry,
+		TimeStarted:     time.Now(),
+	}
+
+	//err := manager.storage.StoreSession("all-sessions", string(sessionID), &se)
+	err := manager.storage.Store("all-sessions", &se)
```
[the remark this answers is not in the document]
Comment (tadovas) [reply core/connection/manager.go]: Another thing. What does "all-sessions" mean ? These table details should be hidden from the caller. Persistence layer should define at take care of that. I liked commented out approach more (with a bit of polishing). manager.store.SaveSession(&se)

File: tequilapi/endpoints/sessions.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func NewSessionsEndpoint(storage storage.Storage) *SessionsEndpoint {
+	return &SessionsEndpoint{
+		storage: storage,
+	}
+}
+
+func (endpoint *SessionsEndpoint) List(resp http.ResponseWriter, request *http.Request, params httprouter.Params) {
+	sessions := []connection.Session{}
```
[the remark this answers is not in the document]
Comment (interro) [reply tequilapi/endpoints/sessions.go]: yes, now there is SessionDTO

File: core/connection/manager.go [hunk 4]
```diff
@@ -61,7 +64,7 @@ type connectionManager struct {
 
 // NewManager creates connection manager with given dependencies
 func NewManager(mysteriumClient server.Client, dialogCreator DialogCreator, promiseIssuerCreator PromiseIssuerCreator,
-	connectionCreator ConnectionCreator, statsKeeper stats.SessionStatsKeeper) *connectionManager {
+	connectionCreator ConnectionCreator, statsKeeper stats.SessionStatsKeeper, storage storage.Storage) *connectionManager {
```
[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager.go]: now it passes SessionRepository from di to connection.NewManager

File: core/connection/manager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager.go]: reworked

File: core/connection/manager.go [hunk 5]
```diff
@@ -270,3 +280,21 @@ func (manager *connectionManager) onStateChanged(state State, sessionID session.
...
+		SessionID:       sessionID,
+		ProviderID:      providerID,
+		ServiceType:     serviceType,
+		ProviderCountry: providerCountry,
+		TimeStarted:     time.Now(),
+	}
+
+	//err := manager.storage.StoreSession("all-sessions", string(sessionID), &se)
+	err := manager.storage.Store("all-sessions", &se)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager.go]: reworked

File: core/connection/manager_test.go [hunk 6]
```diff
@@ -441,3 +452,11 @@ func (fsk *fakeSessionStatsKeeper) GetSessionDuration() time.Duration {
 func (fsk *fakeSessionStatsKeeper) MarkSessionEnd() {
 	fsk.sessionEndMarked = true
 }
+
+type fakeStorage struct{}
+
+func (fs fakeStorage) Store(issuer string, data interface{}) error                         { return nil }
```
[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager_test.go]: created separate file core/storage/storage_fake.go to deduplicate struct

File: core/promise/consumer_test.go [hunk 7]
```diff
@@ -141,3 +141,7 @@ type fakeStorage struct{}
 func (fs *fakeStorage) Store(issuer string, data interface{}) error  { return nil }
 func (fs *fakeStorage) Delete(issuer string, data interface{}) error { return nil }
 func (fs *fakeStorage) Close() error                                 { return nil }
+func (fs *fakeStorage) StoreSession(bucketName string, key string, value interface{}) error {
```
[the remark this answers is not in the document]
Comment (interro) [reply core/promise/consumer_test.go]: created SessionsRepository

File: core/connection/manager.go [hunk 8]
```diff
@@ -270,3 +280,21 @@ func (manager *connectionManager) onStateChanged(state State, sessionID session.
...
+	serviceType string, providerCountry string) error {
+	se := Session{
+		SessionID:       sessionID,
+		ProviderID:      providerID,
+		ServiceType:     serviceType,
+		ProviderCountry: providerCountry,
+		TimeStarted:     time.Now(),
+	}
+
+	//err := manager.storage.StoreSession("all-sessions", string(sessionID), &se)
```
[the remark this answers is not in the document]
Comment (interro) [reply core/connection/manager.go]: reworked

File: core/storage/storage.go [hunk 9]
```diff
@@ -17,9 +17,13 @@
...
 
+import "github.com/asdine/storm"
+
 // Storage stores persistent objects for future usage
 type Storage interface {
 	Store(issuer string, data interface{}) error
+	GetAll(issuer string, data interface{}) error
 	Delete(issuer string, data interface{}) error
 	Close() error
+	GetDB() *storm.DB
```
Comment (vkuznecovas) [inline core/storage/storage.go]: this makes the whole interface rather useless. The storage is now coupled to storm DB. If we change the storage, we'll still have to implement a getdb method that returns a storm db. If you absolutely must have the "GetDB" method, you should then have the method return an interface{} and cast it to the specific DB driver on the consumer side(for instance, tests). In this case I don't think we even need the GetDB

File: core/connection/manager.go [hunk 10]
```diff
@@ -284,3 +292,20 @@ func (manager *connectionManager) onStateChanged(state State, sessionID session.
...
+	providerCountry := connectOptions.Proposal.ServiceDefinition.GetLocation().Country
+	se := Session{
+		SessionID:       connectOptions.SessionID,
+		ProviderID:      connectOptions.ProviderID,
+		ServiceType:     connectOptions.Proposal.ServiceType,
+		ProviderCountry: providerCountry,
+		TimeStarted:     time.Now(),
+	}
+
+	err := manager.sessionStorage.Save(se)
```
Comment (vkuznecovas) [inline core/connection/manager.go]: return manager.sessionStorage.Save(se)

File: core/storage/storage.go (hunk 9 above)
Comment (interro) [reply core/storage/storage.go]: removed GetDB, added needed storage functions to storage interface

**Comment by interro** [comment]
1. Paging can be implemented using Storm queries like: query.Limit(10).Skip(20)

File: core/connection/session_storage.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+package connection
+
+import (
+	"github.com/mysteriumnetwork/node/client/stats"
+	"github.com/mysteriumnetwork/node/core/storage"
+	"github.com/mysteriumnetwork/node/session"
+)
+
+// SessionStorage describes functions for storing session objects
+type SessionStorage interface {
```
Comment (Waldz) [inline core/connection/session_storage.go]: We are agreed to put interfaces to called side, this time it is:
- `connection.Manager` just for `sessionSave` callback
- `sessionsEndpoint` just for `GetAll` callback

File: core/connection/manager.go [hunk 12]
```diff
@@ -270,3 +280,21 @@ func (manager *connectionManager) onStateChanged(state State, sessionID session.
 		manager.status = statusReconnecting()
 	}
 }
+
+func (manager *connectionManager) saveSession(sessionID session.ID, providerID identity.Identity,
```
[the remark this answers is not in the document]
Comment (Waldz) [reply core/connection/manager.go]: I still suggest this to inject smaller interface - just `SaveCallback`

File: core/connection/session_storage.go (hunk 11 above)
Comment (interro) [reply core/connection/session_storage.go]: - Interfaces are now in place.
- [`Save()`](https://github.com/mysteriumnetwork/node/pull/498/files#diff-328f448bf24fd1870fd2370b9c58c0b9R309)

**Comment by Waldz** [comment]
Lgtm

File: core/connection/manager.go [hunk 13]
```diff
@@ -51,13 +52,18 @@ var (
 // consumer identity, provider identity and uses state channel to report state changes
 type ConnectionCreator func(ConnectOptions, StateChannel) (Connection, error)
 
+type sessionStorageSave interface {
```
Comment (soffokl) [inline core/connection/manager.go]: Can we name this interface as `sessionSaver` to follow golang interface naming recommendations?

Comment (interro) [reply core/connection/manager.go]: Will rename it in next PR

**Pull request merged**